### PR TITLE
Add support for Razer Blade 14 (2021)

### LINKED ISSUE #1540: Support for Razer Blade 14
### Device Info

* Name: Razer Blade 14 (2021)
* Product Number: RZ09-0370AW23
* Model Number: RZ09-0370

### Device Information
```
Bus 001 Device 003: ID 1532:0270 Razer USA, Ltd Razer Blade
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x1532 Razer USA, Ltd
  idProduct          0x0270 
  bcdDevice            2.00
  iManufacturer           1 Razer
  iProduct                2 Razer Blade
  iSerial                 0 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0054
    bNumInterfaces          3
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              500mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      61
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     159
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0010  1x 16 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        2
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      2 Mouse
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      94
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
```

```
total 0
drwxr-xr-x 2 root root 180 Jun 28 12:22 .
drwxr-xr-x 4 root root 600 Jun 28 12:22 ..
lrwxrwxrwx 1 root root  10 Jun 28 00:03 usb-Azurewave_Integrated_Camera_200901010001-event-if00 -> ../event13
lrwxrwxrwx 1 root root  10 Jun 28 00:03 usb-Azurewave_Integrated_Camera_200901010001-event-if02 -> ../event14
lrwxrwxrwx 1 root root  10 Jun 28 00:03 usb-Razer_Razer_Blade-event-if01 -> ../event11
lrwxrwxrwx 1 root root   9 Jun 28 00:03 usb-Razer_Razer_Blade-event-kbd -> ../event5
lrwxrwxrwx 1 root root   9 Jun 28 00:03 usb-Razer_Razer_Blade-if01-event-kbd -> ../event8
lrwxrwxrwx 1 root root  10 Jun 28 00:03 usb-Razer_Razer_Blade-if02-event-mouse -> ../event12
lrwxrwxrwx 1 root root   9 Jun 28 00:03 usb-Razer_Razer_Blade-if02-mouse -> ../mouse2
```

### Synapse Screenshots
screenshot-1
screenshot-2
screenshot-3

### PR DESCRIPTION
**Pull request by beeaniebee**
Works as expected, no issues found.
fixes #1540

### CODE REVIEW

**Comment by beeaniebee** [comment]
Check failed because I was unable to find the image link for the device. Added the variable so that it easily be added by a moderator if found.

**Review by mrmetaverse — APPROVED** [review]
Looks good to me, can we get this merged?

File: README.md [hunk 1]
```diff
@@ -76,6 +76,7 @@ The devices below are fully feature supported by OpenRazer, which means all avai
 | Razer Book 13 (2020)                          |  1532:026A  |
 | Razer Blade 15 Advanced (Early 2021)          |  1532:026D  |
 | Razer Blade 15 Advanced (Mid 2021)            |  1532:0276  |
+| Razer Blade 14 (2021)                         |  1532:2070  |
```
Comment (z3ntu) [inline README.md]: Should be `1532:0270` and move it one line up to keep it sorted

File: daemon/openrazer_daemon/hardware/keyboards.py [hunk 2]
```diff
@@ -1341,6 +1341,24 @@ class RazerBlade15Advanced2021(_RippleKeyboard):
...
+    HAS_MATRIX = True
+    MATRIX_DIMS = [6, 16]
+    METHODS = ['get_device_type_keyboard', 'get_logo_active', 'set_logo_active', 'set_wave_effect', 'set_static_effect', 'set_spectrum_effect',
+               'set_reactive_effect', 'set_none_effect', 'set_breath_random_effect', 'set_breath_single_effect',
+               'set_breath_dual_effect', 'set_custom_effect', 'set_key_row',
+               'set_starlight_random_effect', 'set_starlight_single_effect', 'set_starlight_dual_effect',
+               'set_ripple_effect', 'set_ripple_effect_random_colour']
+
+
+    DEVICE_IMAGE = ""
```
Comment (z3ntu) [inline daemon/openrazer_daemon/hardware/keyboards.py]: How about https://dl.razerzone.com/Images/blade15advanced%282021%29/Razer%20Blade%2015%20Advanced5_00001.png for the picture?

File: driver/razerkbd_driver.h [hunk 3]
```diff
@@ -79,6 +79,7 @@
 #define USB_DEVICE_ID_RAZER_BOOK_2020 0x026A
 #define USB_DEVICE_ID_RAZER_BLADE_15_ADV_EARLY_2021 0x026D
 #define USB_DEVICE_ID_RAZER_BLADE_15_ADV_MID_2021 0x0276
+#define USB_DEVICE_ID_RAZER_BLADE_14_2021 0x0270
```
Comment (z3ntu) [inline driver/razerkbd_driver.h:81]: Move it one up to keep it sorted

File: install_files/appstream/io.github.openrazer.openrazer.metainfo.xml [hunk 4]
```diff
@@ -141,6 +141,7 @@
     <modalias>usb:v1532p026Ad*</modalias>
     <modalias>usb:v1532p026Dd*</modalias>
     <modalias>usb:v1532p0276d*</modalias>
+    <modalias>usb:v1532p0270d*</modalias>
```
Comment (z3ntu) [inline install_files/appstream/io.github.openrazer.openrazer.metainfo.xml:143]: Regenerate this file with `scripts/generate_appstream_file.sh`

File: install_files/udev/99-razer.rules [hunk 5]
```diff
@@ -10,7 +10,7 @@ ATTRS{idProduct}=="0013|0016|0020|0024|0025|002e|002f|0032|0034|0036|0037|0038|0
     ENV{ID_RAZER_CHROMA}="1", ENV{RAZER_DRIVER}="razermouse"
 
 # Keyboards
-ATTRS{idProduct}=="010d|010e|010f|0111|0113|011a|011b|011c|0201|0202|0203|0204|0205|0207|0208|0209|020f|0210|0211|0214|0216|0217|021a|021e|021f|0220|0221|0224|0225|0226|0227|0228|022a|022b|022c|022d|022f|0232|0233|0234|0235|0237|0239|023a|023b|023f|0240|0241|0243|0245|0246|024a|024c|024d|0252|0253|0255|0257|0259|025d|025e|0266|026a|026d|0276|0a24", \
+ATTRS{idProduct}=="010d|010e|010f|0111|0113|011a|011b|011c|0201|0202|0203|0204|0205|0207|0208|0209|020f|0210|0211|0214|0216|0217|021a|021e|021f|0220|0221|0224|0225|0226|0227|0228|022a|022b|022c|022d|022f|0232|0233|0234|0235|0237|0239|023a|023b|023f|0240|0241|0243|0245|0246|024a|024c|024d|0252|0253|0255|0257|0259|025d|025e|0266|026a|026d|0276|0270|0a24", \
```
Comment (z3ntu) [inline install_files/udev/99-razer.rules]: move the pid of your device one to the front to keep it sorted

**Review by z3ntu — CHANGES_REQUESTED** [review]
Some minor style problems. Also note the CI output:

```
+ ./scripts/ci/check-autopep8-formatting.sh
[!] ./daemon/openrazer_daemon/hardware/keyboards.py does not respect the agreed coding style.

You can run ./scripts/format_source.sh to automatically format those files.
```

File: daemon/openrazer_daemon/hardware/keyboards.py (hunk 2 above)
Comment (johnhout) [reply daemon/openrazer_daemon/hardware/keyboards.py]: @z3ntu  thats the blade 15 :smile: should be:
https://assets2.razerzone.com/images/og-image/razer-blade-14-og-image-1200x630.jpg

File: README.md (hunk 1 above)
Comment (johnhout) [reply README.md]: @z3ntu  @BeanieBen9990  Made a pull request for this change.

File: daemon/openrazer_daemon/hardware/keyboards.py (hunk 2 above)
Comment (johnhout) [reply daemon/openrazer_daemon/hardware/keyboards.py]: @z3ntu  @BeanieBen9990  Made a pull request for this change.

File: driver/razerkbd_driver.h (hunk 3 above)
Comment (johnhout) [reply driver/razerkbd_driver.h:81]: @z3ntu  @BeanieBen9990  Made a pull request for this change.

File: install_files/appstream/io.github.openrazer.openrazer.metainfo.xml (hunk 4 above)
Comment (johnhout) [reply install_files/appstream/io.github.openrazer.openrazer.metainfo.xml:143]: @z3ntu  @BeanieBen9990  Made a pull request for this change.

File: install_files/udev/99-razer.rules (hunk 5 above)
Comment (johnhout) [reply install_files/udev/99-razer.rules]: @z3ntu  @BeanieBen9990  Made a pull request for this change.

**Comment by beeaniebee** [comment]
Merged those pull requests into mine. Happy to fix these issues w the PR today, once I've got some time. Been very busy but should be totally ready very soon, hopefully by the end of the day.

**Comment by johnhout** [comment]
Added a pull-request for the code style fixes as sugested by @z3ntu. Should resolve the failing check

**Comment by beeaniebee** [comment]
Merged that too. Will still look back on everything later today

File: README.md (hunk 1 above)
Comment (johnhout) [reply README.md]: @z3ntu the formatting as suggested has been merged :smile:

File: daemon/openrazer_daemon/hardware/keyboards.py [hunk 6]
```diff
@@ -87,7 +87,7 @@ class RazerNostromo(_RazerDeviceBrightnessSuspend):
                # ?
                'keypad_get_mode_modifier', 'keypad_set_mode_modifier']
 
-    DEVICE_IMAGE = "https://assets.razerzone.com/eeimages/support/products/564/564_tartarus_classic.png"
+    DEVICE_IMAGE = "https://assets.razerzone.com/eeimages/support/products/564/Blade14-Piper-FHD.png"
```
Comment (ronplanken) [inline daemon/openrazer_daemon/hardware/keyboards.py]: This change seems to have been snuck in after running the linter. The link is broken and it seems to be unintended? This should probably be reverted.

**Review by ronplanken — CHANGES_REQUESTED** [review]

File: daemon/openrazer_daemon/hardware/keyboards.py (hunk 6 above)
Comment (beeaniebee) [reply daemon/openrazer_daemon/hardware/keyboards.py]: If you make a pull request on my fork I'll merge it quickly, otherwise I'll make the change later

Comment (johnhout) [reply daemon/openrazer_daemon/hardware/keyboards.py]: @BeanieBen9990 Done 👍

**Comment by johnhout** [comment]
@z3ntu is this ready to get merged?

**Comment by johnhout** [comment]
@z3ntu bump

**Comment by ttaaxxiill** [comment]
> Works as expected, no issues found.
> fixes #1540

Please note, I'm seeing 1532:**0270** listed on my machine's lsusb output, but throughout this patch I see it referred to as **2720**

**Comment by beeaniebee** [comment]
> > Works as expected, no issues found.
> 
> > fixes #1540
> 
> 
> 
> Please note, I'm seeing 1532:**0270** listed on my machine's lsusb output, but throughout this patch I see it referred to as **2720**

I can't see what you're seeing. Throughout the patch, the number is **0270**

**Comment by ttaaxxiill** [comment]
> > > Works as expected, no issues found.
> > 
> > 
> > > fixes #1540
> > 
> > 
> > Please note, I'm seeing 1532:**0270** listed on my machine's lsusb output, but throughout this patch I see it referred to as **2720**
> 
> I can't see what you're seeing. Throughout the patch, the number is **0270**

Apologies... it was only in the conversation. Any chance of this being integrated soon... my touchpad is deadzo.

**Comment by beeaniebee** [comment]
> my touchpad is deadzo.

This doesn't affect the touchpad at all. The touchpad should work just fine with or without it.

**Comment by ttaaxxiill** [comment]
> > my touchpad is deadzo.
> 
> This doesn't affect the touchpad at all. The touchpad should work just fine with or without it.

Well, shit.

**Comment by BentonEdmondson** [comment]
Lel

**Comment by BentonEdmondson** [comment]
@z3ntu I beg of you my lord

**Comment by johnhout** [comment]
@z3ntu bump :)

**Comment by bskrt** [comment]
@z3ntu Ping!

**Comment by garybowers** [comment]
Any update on this? PR has been open a while.

**Comment by mrmetaverse** [comment]
Right, looks like we're just waiting on a merge. 

Bump

**Review by z3ntu — APPROVED** [review]
LGTM, thanks!

**Pull request merged**